# Taxi, Lyft, Uber, or Car Rental
## Figuring Out Your Costs



One decides to take a trip to Seattle. There are many options for getting around the city. If one does not mind sitting next to random strangers on a crowded bus, public transportation is an option. Other options include getting around by taxi, Lyft, or providing your own transportatin by renting a car. Which options are the cheapest? It would probably depend on how much you need to get around the city. Trips covering large amounts of distance with lengthy trip times might add up quickly to where renting a car may be the more viable option. To help calculate the cost of getting around using the various methods excluding public transportation, some functions will be set up to help figure out the total costs. The goal of this project is to assume someone wants to figure out how much they will spend depending on which option they choose. They would be able to approximate their costs by knowing how far they will travel and plugging in various time ranges.
Units used in this project:
* Distance: Miles
* Time: minutes
* Currency: USD


Option 1. Standard taxi service.  The three main costs to consider when using a taxi:
1. Drop charge
2. Distance charge
3. Time charge


In [1]:
import random

In [2]:
def taxi_cost(distance, time, drop = 2.60): 
    distance_charge = distance * 2.70
    time_charge = time * 0.50
    total_taxi_cost = distance_charge + time_charge +  drop     
    return round(total_taxi_cost, 2)

print(taxi_cost(3.5, 5.4))
# Check to see that the function works properly. 
# Distance traveled is 3.5 miles
# Total time is 5.4 minutes. 
print(round((3.5 * 2.70) + (5.4 * 0.50) + 2.60, 2))
# Looks like the function works correctly

14.75
14.75


Option 2. Lyft and factors to consider in costs
  
1. Initial cost
2. Service fee
3. Distance charge
4. Time charge
5. City tax
6. Mininum fare
7. Maximum fare

In [3]:
def lyft_cost(distance, time, initial = 1.53, fee = 2.00, tax = 0.24):
    distance_charge = distance * 1.60
    time_charge = time * 0.27
    total_lyft_cost = distance_charge + time_charge + initial + fee + tax
    minimum_fare = 3.50
    maximum_fare = 400.00
    if total_lyft_cost <= 3.50:
        price = minimum_fare
    elif total_lyft_cost >= 400.00:
        price = maximum_fare
    else:
        price = total_lyft_cost
    return round(price,2)

# Testing out the lyft_cost function. 
print(lyft_cost(.5, 1, fee = .50))
print((1.60 * .5) + (1 * 0.27) + 1.53 + .50 + 0.24)
print('\n')
# Result = 3.34 which is less than 3.5 and the function ouputs value of 3.5 which checks
# Going to check if nothing greater than 400.00 gets returned.
print(lyft_cost(250, 240))
print((250 * 1.60) + (240 * 0.27) + 1.53 + 2.00 + 0.24)
print('\n')
# 468.57 > 400.00 It looks like function is working correctly. Peform 3rd check below.
print(lyft_cost(15, 30))
print(round((15 * 1.60) + (30 * 0.27) + 1.53 + 2.00 + .24, 2))
# Function checks

3.5
3.34


400.0
468.57


35.87
35.87


Option 3. Uber and factors to consider in costs:
1. Base fare
2. Time charge
3. Distance charge
4. Service fee


In [4]:
def uber_cost(distance, time, model):
    if model == 'UberX':
        distance_charge = distance  * 1.35
        time_charge = time * 0.24
        cost = distance_charge + time_charge + 1.35 + 1.30
    elif model == 'UberXL':
        distance_charge = distance * 2.75
        time_charge = time * 0.30
        cost = distance_charge + time_charge + 3.00 + 1.30
    elif model == 'UberBLACK':
        distance_charge = distance * 3.75
        time_charge = time * 0.35
        cost = distance_charge + time_charge + 7.00
    elif model == 'UberSUV':
        distance_charge = distance * 4.20
        time_charge = time * 0.40
        cost = distance_charge + time_charge + 14.00
    elif model =='UberSELECT':
        distance_charge = distance * 2.75
        time_charge = time * 0.30
        cost = distance_charge + time_charge + 4.00 + 1.20
    return round(cost, 2)
        

Running a series of tests to make sure the 'uber_cost' function is working correctly. Will use default values of 10 miles for distance and 30 minutes for time.


In [5]:
print(uber_cost(10, 30, 'UberX'))
print(uber_cost(10, 30, 'UberXL'))
print(uber_cost(10, 30, 'UberBLACK'))
print(uber_cost(10, 30, 'UberSUV'))
print(uber_cost(10, 30, 'UberSELECT'))
print('\n')
# Output with different values which indicates function is working properly
# Will do a random check to make sure calculations in function are working properly
# Will use UberBLACK which should come out to 55.0 
print((10 * 3.75) + (30 * 0.35) + 7.00)
# Looks like the correct calculations within function are properly working

23.35
40.8
55.0
68.0
41.7


55.0


Option 4. Renting your own vehicle

In [6]:
def car_rental(car, num_days):
    if car == 'standard':
        cost = num_days * 24.00
    elif car == 'full-size':
        cost = num_days * 30.00
    elif car == 'SUV':
        cost = num_days * 34.00
    elif car == 'minivan':
        cost = num_days * 43.00
    return round(cost, 2)
print(car_rental('standard', 2))
print(car_rental('SUV', 3))
# The 'car_rental' function appears to be error free

48.0
102.0


Example scenario:

Maverick Scorzeeny plans to take a seven-day trip to Seattle and will be traveling to and from a convention about 6.4 miles each way (12.8 miles roundtrip) from his hotel. Due to the fact he will be commuting during heavy traffic both his outbound and inbound trips will run about 30 minutes each! He will only have to make this commute once a day. How much can he plan to spend using the various options? Will he opt for Uber or Lyft but may have to rely on long wait times, or will he opt for taking the taxi due to the quick response times? Could renting a car be the more viable option? What will his costs be using the various methods over the seven-day period?

In [7]:
taxi_price = taxi_cost(6.4, 30)
print('It will cost an additional $' + str(round(((taxi_price * 7) *2 ),2)) + ' to his trip taking the taxi!')
print('\n') # Spacing between each entry for less cluttered appearance 
lyft_price = lyft_cost(6.4, 30)
print('It will cost an additional $' + str(round(((lyft_price * 7) * 2),2)) + ' to his trip using lyft.')
print('\n')
uberx_cost = uber_cost(6.4, 30, 'UberX')
print('It will cost an additional $' +str(round(((uberx_cost * 7) * 2), 2)) + ' to his trip using UberX.')
print('\n')
uber_black_cost = uber_cost(6.4, 30, 'UberBLACK')
print('It will cost an additional $' + str(round(((uber_black_cost * 7) * 2),2)) + ' to his trip using UberBLACK!')
print('\n')
standard_rental = car_rental('standard', 7)
print('Renting a standard economy car will cost him an additional $' + str(round(standard_rental)) + ' exluding gas.')

It will cost an additional $488.32 to his trip taking the taxi!


It will cost an additional $309.54 to his trip using lyft.


It will cost an additional $258.86 to his trip using UberX.


It will cost an additional $581.0 to his trip using UberBLACK!


Renting a standard economy car will cost him an additional $168 exluding gas.


In [8]:
summary = '''With so many choices to make when it comes to which mode or carrier to use for transportation
one can weigh the options between quick and reliable service that may cost a bit more, or opt for the 
cheapest mode regardless of wait times. At ${0} each trip for {1}, that should make less of a dent on his wallet. However, 
one might save even more, much more, if they are willing to do the driving themselves.'''.format(uberx_cost, 'UberX')

In [9]:
print(summary)

With so many choices to make when it comes to which mode or carrier to use for transportation
one can weigh the options between quick and reliable service that may cost a bit more, or opt for the 
cheapest mode regardless of wait times. At $18.49 each trip for UberX, that should make less of a dent on his wallet. However, 
one might save even more, much more, if they are willing to do the driving themselves.


Suppose Maverick arrives to Seattle and there are no available car rentals and Uber is out of service, how much can Maverick expect to save over the course of seven days using Lyft? The following python code below will attempt to assist Maverick in calculating his savings!

In [10]:
def taxi_vs_lyft(distance, time):
    taxi =  (taxi_cost(distance, time)) * 2 
    lyft = (lyft_cost(distance, time)) * 2
    if taxi < lyft:
        return ('It looks like $' + str(round((lyft * 7) - (taxi * 7), 2))  + ' will be saved over a seven-day period using the infamous yellow cab!')
    elif lyft < taxi:
        return ('It looks like $' + str(round((taxi * 7) - (lyft * 7), 2))  + ' will be saved over a seven-day period using Lyft!')
savings = taxi_vs_lyft(6.4, 30)
print(savings)

It looks like $178.78 will be saved over a seven-day period using Lyft!


Over the course of a seven-day perioud, Maverick will save $178.78!
 

Creating a bot to help assist in determining
total trip cost based on ones chosen transportation preferences assuming they do not opt to rent a vehicle. However, let us assume it is a real busy day in the city and  the user will most likely be in a hurry. If the user opts to use a taxi service, he/she can rest assure the wait time will be less than 16 minutes. It could be as long as a 45 minute wait for Lyft or Uber. For those opting for Uber or Lyft the bot will attempt to give a random estimated wait time.

In [11]:
destination = input('Where do you want to go? ')

Where do you want to go? Northgate


In [12]:
miles = float(input('Enter approximate distance: '))


Enter approximate distance: 5.7


In [13]:
duration = float(input('That will take about: '))  

That will take about: 18.4


In [14]:
option = input('Which mode of transportation do you want to use? ')

Which mode of transportation do you want to use? Uber


In [15]:
def message(d = destination, m = miles, t = duration, o = option): # Parameters set to default values from input values above
    a_string = 'So, you want to go to ' + d + '?' + ' That is ' + str(m) + ' miles away.'
    b_string = ' With traffic that will take around ' + str(t) + ' minutes.'
    c_string = ' I see you will like to use ' + o +'.'
    d_string = " It's really busy today. So let's check on your approximated wait time."
   
    
     
    return a_string + b_string + c_string + d_string
    
message()

"So, you want to go to Northgate? That is 5.7 miles away. With traffic that will take around 18.4 minutes. I see you will like to use Uber. It's really busy today. So let's check on your approximated wait time."

User will then use the function below to receive a random wait time put out by the bot. 

In [16]:
def wait_time():
    print("Estimating your wait time.")
    time = random.randint(0, 45)
    print('It looks like your waiting time is ' +str(time) + ' minutes!')
    if time <= 15:
        print('YES!!' + ' I will wait the ' + str(time) + ' minutes')
    else:
        print(str(time) + ' minutes!!' + ' I will take the taxi instead!')
    
wait_time()

Estimating your wait time.
It looks like your waiting time is 20 minutes!
20 minutes!! I will take the taxi instead!


Having weighed ones options and considering wait times. One would use the 'miles' and 'duration' variables declared above as arguments into the various functions delcared above to get an estimated price for each of the transportation options.

In [17]:
taxi_chosen = taxi_cost(miles, duration)  
print('Using the taxi will cost you about $' + str(taxi_chosen) +'.') 
lyft_chosen = lyft_cost(miles, duration)
print('Using Lyft cost you around $' + str(lyft_chosen) +'.')
uber_x_chosen = uber_cost(miles, duration, model = 'UberX')
print('The UberX option will cost you around $' + str(uber_x_chosen) + '.')
uber_xl_chosen = uber_cost(miles, duration, model = 'UberXL')
print('The UberXL option will cost you around $' + str(uber_xl_chosen) + '.')
uber_black_chosen = uber_cost(miles, duration, model = 'UberBLACK')
print('The UberBLACK option will cost you around $' + str(uber_black_chosen) + '.')
uber_suv_chosen = uber_cost(miles, duration, model = 'UberSUV')
print('The UberSUV option will cost you around $' + str(uber_suv_chosen) + '.')
uber_select_chosen = uber_cost(miles, duration, model = 'UberSELECT')
print('The UberSELECT option will cost you around $' + str(uber_select_chosen) + '.')


Using the taxi will cost you about $27.19.
Using Lyft cost you around $17.86.
The UberX option will cost you around $14.76.
The UberXL option will cost you around $25.5.
The UberBLACK option will cost you around $34.81.
The UberSUV option will cost you around $45.3.
The UberSELECT option will cost you around $26.39.


Given the destination, distance, and approximate trip duration input given by the user, the approximated cost of the trip can be shown to the user. 

This project utilized some control flow within some of the defined functions above. String concatenation along with converting integer or float values to strings to enable string concatenation in print statements was also utilized. This project also consisted of using some input statements and assigning them to variables. With more options to seemingly choose from when it comes to getting around the city, it can sure be beneficial to at least know before hand about how much a trip will cost!


    
    
